File: UI/shared-types/src/common/api-response.ts
/**
 * Shared API response types — mirrors the backend ResponseMapperInterceptor output.
 * Source of truth: API/src/common/response/response-mapper.interceptor.ts
 */

// ── Pagination ───────────────────────────────────────────────────────────────

export interface PaginationMeta {
  page: number;
  limit: number;
  total: number;
  totalPages: number;
  hasNext: boolean;
  /** Backend field: hasPrevious (from ApiResponse.paginated() and api-response.interface.ts) */
  hasPrevious: boolean;
}

// ── Standard API envelope ────────────────────────────────────────────────────

/**
 * Standard response wrapper from the NestJS ResponseMapperInterceptor.
 * Every successful backend response is shaped like this.
 */
export interface ApiResponse<T> {
  success: boolean;
  statusCode: number;
  message: string;
  data: T;
  meta?: PaginationMeta;
  timestamp: string;
  path: string;
  requestId: string;
}

// ── Error shape ──────────────────────────────────────────────────────────────

export interface ApiErrorBody {
  code: string;
  message: string;
  details?: Record<string, unknown> | null;
  suggestion?: string;
  documentationUrl?: string;
}

export interface ApiErrorResponse {
  success: false;
  statusCode: number;
  message: string;
  error: ApiErrorBody;
  timestamp: string;
  path: string;
  requestId: string;
}

// ── Paginated data wrapper ───────────────────────────────────────────────────

export interface Paginated<T> {
  data: T[];
  total: number;
  page: number;
  limit: number;
  totalPages: number;
  hasNext: boolean;
  hasPrevious: boolean;
}

// ── Type guards ──────────────────────────────────────────────────────────────

export const isApiSuccess = <T>(res: ApiResponse<T> | ApiErrorResponse): res is ApiResponse<T> =>
  res.success === true;

export const isApiError = (res: ApiResponse<unknown> | ApiErrorResponse): res is ApiErrorResponse =>
  res.success === false;

// ── Result alias (functional-style, mirrors backend ResultType<T>) ───────────

export type ResultType<T> =
  | { success: true; data: T }
  | { success: false; error: ApiErrorBody };

export const isOk = <T>(result: ResultType<T>): result is { success: true; data: T } =>
  result.success === true;

export const isErr = <T>(result: ResultType<T>): result is { success: false; error: ApiErrorBody } =>
  result.success === false;

File: mobile/marketplace/lib/core/network/api_response.dart
class ApiResponse<T> {
  final bool success;
  final T? data;
  final String? message;
  final String? errorCode;

  const ApiResponse({
    required this.success,
    this.data,
    this.message,
    this.errorCode,
  });

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic) fromJsonT,
  ) {
    return ApiResponse(
      success: json['success'] as bool? ?? false,
      data: json['data'] != null ? fromJsonT(json['data']) : null,
      message: json['message'] as String?,
      errorCode: json['errorCode'] as String?,
    );
  }
}

class PaginatedResponse<T> {
  final List<T> data;
  final int page;
  final int limit;
  final int total;
  final bool hasMore;

  const PaginatedResponse({
    required this.data,
    required this.page,
    required this.limit,
    required this.total,
    required this.hasMore,
  });

  factory PaginatedResponse.fromJson(
    Map<String, dynamic> json,
    T Function(dynamic) fromJsonT,
  ) {
    final rawData = json['data'];
    final List<dynamic> items = rawData is List
        ? rawData
        : (rawData is Map && rawData['data'] is List ? rawData['data'] : []);
    final meta = rawData is Map ? rawData['meta'] : json['meta'];

    return PaginatedResponse(
      data: items.map(fromJsonT).toList(),
      page: meta?['page'] as int? ?? 1,
      limit: meta?['limit'] as int? ?? 20,
      total: meta?['total'] as int? ?? 0,
      hasMore: meta?['hasMore'] as bool? ?? false,
    );
  }
}
